## Modified in Jan2021:
## better stratgey: 

## necessary pre-processing steps:
1. subset donor vcf files to fast loading
`bcftools view merged.chr.vcf.gz -R cellSNP_out/cellSNP.cells.vcf.gz -Oz -o sub.merged.vcf.gz`

2. doulbe check whether the output cellSNP.cells.vcf.gz has the same forrmat of chrom as the donor vcf file after finished cellSNP

`bcftools query -f "%CHROM\n" cellSNP.cells.vcf.gz | less`
`bcftools query -f "%CHROM\n" donor.vcf.gz | less`

if they are not consistent, need to change

`cp ../compare-variants/chr_name_conv.txt ./`

```
for file in folder1 folder2
do
bcftools annotate --threads 56 --rename-chrs chr_name_conv.txt  NIHPlR1/cellSNP/${file}.vcf.gz -Oz -o NIHPlR1/cellSNP/${file}.chr.vcf.gz
bcftools index --threads 56 NIHPlR1/cellSNP/${file}.chr.vcf.gz
done
```

### adjust runing parameters according file size
1. separate cellsnp and vireo steps: cellsnp takes longer time but don't need as much mem; usually less than 20g and 24hours depending on the file. so can set 

`swarm -f cellsnp_2021.swarm -g 40 -t 20 --time 48:00:00 --module cellsnp/0.1.7`

`swarm -f vireo_2021.swarm -g 240 -t 20 --time 8:00:00 --module vireosnp/0.3.2`

2. vireo takes much more memory, sometimes 120g enough, but sometime need 240g,time may be 1 hour or a few hours
also better first try one in sinteractive instead of swarm, this way, more straigtforward for error

`sinteractive --mem=240g --cpus-per-task=20`

## common errors:
1. the exist of cellSNP or vireo_out from last fail run would cause new error in swarm: change linux scripts of mkdir using if
`[ ! -d "cellSNP_out" ] && mkdir cellSNP_out/ && echo "created the cellSNP_out folder"`
`[ ! -d "vireo_out" ] && mkdir vireo_out/ && echo "created the vireo_out folder"`

## change the structure of foldeer:
if there are multiple folders need to be run, the change on sh, swarm, or donor vcf would need change all the copies in every folder. This is not a good way. So put all the common/shared folders in one path, then only change this copy.

also make sure that the files inside all sampels folders share the file names

- current/SAMPLE1/"sample specific files"
- current/"shared files and sciprts

make sure the correct folder: sub.donor.vcf is inside the sample folder because it is subset according to the cellSNP_out/cellSNP.cells.vcf.gz

`bcftools view merged.chr.vcf.gz -R Sample/cellSNP_out/cellSNP.cells.vcf.gz -Oz -o Sample/sub.merged.vcf.gz`

cellSNP_out and vireo_out are automatically generated by scripts

CURRENT_PATH
│
── SAMPLE1
│   ├── barcodes.tsv.gz
│   ├── cellSNP_out
│   │   ├── cellSNP.base.vcf.gz
│   │   ├── cellSNP.cells.vcf.gz
│   │   ├── cellSNP.samples.tsv
│   │   ├── cellSNP.tag.AD.mtx
│   │   ├── cellSNP.tag.DP.mtx
│   │   └── cellSNP.tag.OTH.mtx
│   ├── possorted_genome_bam.bam
│   ├── possorted_genome_bam.bam.bai
│   ├── sub.UNHS_to_use_7s.vcf.gz # subseted donor genotype file 
│   └── vireo_out
│       ├── donor_ids.tsv
│       ├── fig_GT_distance_estimated.pdf
│       ├── _log.txt
│       ├── prob_doublet.tsv.gz
│       ├── prob_singlet.tsv.gz
│       └── summary.tsv
└── cellsnp.sh
└── cellsnp.swarm
└── vireo.sh
└── vireo.swarm
└── UNHS_to_use_7s.vcf.gz # donor genotype file without subset
└── genome1K.phase3.SNP_AF5e2.chr1toX.hg38.vcf.gz # reference genome's genotype file

For this way have to give the sample folder name as the argument of bash cellsnp.sh inside the swarm file

### cellsnp_2021_up.swarm
bash cellsnp_2021_up.sh NIHPlR1
bash cellsnp_2021_up.sh NIHPlR2
bash cellsnp_2021_up.sh NIHPlR3

### vireo_2021_up.swarm
bash vireo_2021_up.sh NIHPlR1
bash vireo_2021_up.sh NIHPlR2
bash vireo_2021_up.sh NIHPlR3


